chore(clip): update the clip example to make it compatible with HF format

mzbac opened the pull request:
Update the clip model format to HF format to make it easier for later use as a vision tower in the llava model.
close https://github.com/ml-explore/mlx-examples/issues/463

![image](https://github.com/ml-explore/mlx-examples/assets/7523197/b2c1de34-6d37-4fdc-90fa-8107e19aa6e8)

awni commented on clip/requirements.txt:
```diff
@@ -3,4 +3,5 @@ numpy
-Pillow
\ No newline at end of file
+Pillow
+mlx-lm
```
I would prefer not to make `mlx-lm` a dep of this package. It might make things a little too brittle

awni replied:
It's just for `save_weights`, so I would just copy a possibly simplified version of that here.

mzbac replied:
yeah, make sense

awni commented on clip/convert.py:
```diff
@@ -86,7 +49,13 @@ def should_keep_weight(key: str):
-
+    parser.add_argument(
+        "--dtype",
+        help="dtype for loading the torch model and input for quantization or saving the converted model. "
+        "The original weights are stored in bfloat16.",
```
```suggestion
        help="The data type to save the converted model. "
```

I didn't get the ocmment about `bfloat16`?

mzbac replied:
Sorry, it was copied from the mlx-lm converter.py and I forgot to update the description. I will clean up the description.

awni commented on clip/model.py:
```diff
@@ -75,50 +80,179 @@ def clip_loss(logits: mx.array) -> mx.array:
+        return y
+
+
+class CLIPAttention(nn.Module):
```
nit, I don't think we need to prepend `CLIP` to everything in this file, maybe just the layers that are panned to export. Makes it a little verbose..

awni commented on clip/model.py:
```diff
@@ -75,50 +80,179 @@ def clip_loss(logits: mx.array) -> mx.array:
+        return self.out_proj(values_hat)
+
+
+class CLIPMLP(nn.Module):
```
Same comment here.

mzbac commented:
@awni, thanks for helping with reviewing the PR. I have addressed all the comments, it should be ready to merge now. Please let me know if there are any other changes needed.

mzbac commented on clip/model.py:
```diff
@@ -170,25 +299,50 @@ def _embed(self, x: mx.array) -> mx.array:
+    def __call__(
+        self,
+        x: mx.array,
+        output_hidden_states: Optional[bool] = None,
```
@awni added support for output_hidden_states in the ClipVisionModel, which will be used on the llava model. Also updated the test to cover this. Let me know if you want me to add this in a separate PR.

awni replied:
I think it's fine to include it here!

mzbac commented on clip/test.py:
```diff
@@ -86,12 +86,14 @@ def test_vision_encoder(self):
             expected_last_hidden = expected_out.last_hidden_state.numpy()
             expected_pooler_output = expected_out.pooler_output.numpy()
-
+            expected_hidden_states = [hs.numpy() for hs in expected_out.hidden_states]
```
Updated the test to cover the output_hidden_states option.

awni commented:
@mzbac let me know when I should check this again

mzbac commented:
@awni, this is ready for merging. Once this is merged, I will work with @nkasmanoff on https://github.com/ml-explore/mlx-examples/pull/461

awni commented on clip/model.py:
```diff
@@ -450,5 +406,16 @@ def from_pretrained(path: str):
+            if "position_ids" in k:
+                # Remove unused position_ids
+                continue
+            elif "patch_embedding.weight" in k:
```
👍

awni commented on clip/model.py:
```diff
@@ -170,25 +255,50 @@ def _embed(self, x: mx.array) -> mx.array:
+        encoder_states = () if output_hidden_states else None
+
+        for l in self.encoder.layers:
+            if output_hidden_states:
+                encoder_states = encoder_states + (x,)
             x = l(x, mask=None)
```
```suggestion
            x = l(x, mask=None)
            if output_hidden_states:
                encoder_states = encoder_states + (x,)
```

awni commented on clip/model.py:
```diff
@@ -170,25 +255,50 @@ def _embed(self, x: mx.array) -> mx.array:
+
+        encoder_states = () if output_hidden_states else None
+
+        for l in self.encoder.layers:
+            if output_hidden_states:
+                encoder_states = encoder_states + (x,)
             x = l(x, mask=None)
 
+        if output_hidden_states:
+            encoder_states = encoder_states + (x,)
```
```suggestion
        for l in self.encoder.layers:
            x = l(x, mask=None)
            if output_hidden_states:
                encoder_states = encoder_states + (x,)

```

awni reviewed (approved):
Looks great, thank you!!

awni commented:
Thanks!!
